docs/architecture/clients/dependency-injection.md

```diff
@@ -0,0 +1,219 @@
+implements the abstract interface, or that the `deps` array matches the constructor parameters.
+
+We provide a helper function called `safeProvider`, which acts as a wrapper around each provider
+entry and provides compile-time type checking of your configuration. For example (continuing with
+the example above):
+
+```ts
+@NgModule({
+  providers: [
+    safeProvider({
+      provide: MyService,
+      useClass: DefaultMyService,
+      deps: [DepA, DepB],
+    }),
+    safeProvider({
+      provide: SomeAngularService,
+      useClass: DefaultSomeAngularService,
+      useAngularDecorators: true, // this tells safeProvider to not require a deps array
+    }),
+  ],
+})
+export class MyServicesModule {}
+```
+
+:::note
+
+The `useAngularDecorators` option is not type safe - it is your responsibility to ensure that the
+class actually uses the `@Injectable` decorator.
+
+:::
+
+Our primary service modules require the use of `safeProvider`, otherwise you will receive type
+errors. However, you should use `safeProvider` wherever you are configuring an Angular provider to
+benefit from its type checking.
+
+### Angular root module
+
+Services that will be used globally and which don't have an abstract interface can be registered
+using the `provideIn` property. This avoids having to create and manage service modules. When
+exporting services through feature modules, it can be tricky to manage their lifetime, and it's easy
+to end up with instances of the service being created.
+
+```ts
+@Injectable({
+  provideIn: "root",
+})
+export class MyService {}
+```
+
+### Angular components
+
+Services that do are not stateful and/or are only supposed to be used locally to extract complex
+logic from a component can be injected directly into standalone components, completely bypassing
+modules.
+
+```ts
+@Component({
+  providers: [
+    // This example assumes that MyService uses @Injectable decorators and does not have an abstract interface
+    safeProvider(MyService),
+  ],
+  standalone: true,
+})
+export class MyComponent {}
+```
+
+### Non-Angular contexts
+
+Non-Angular contexts do not use a DI framework and must manually instantiate their dependencies **in
+the correct order**. Instantiating dependencies out of order may result in null values being
+injected into services, causing runtime errors.
+
+## DI by client
+
+### Shared Angular DI
+
+`libs/angular` contains `jslib-services.module.ts`, a services module which registers common
+dependencies used across Angular Typescript clients (web, browser and desktop).
+
+A client may override this DI if a client-specific configuration is required.
+
+### Web
+
+`core.module.ts` contains the main services module for the web client. It imports
+`JslibServicesModule`.
+
+However, the web module heavily uses [feature modules](https://angular.io/guide/feature-modules) to
+divide its code by feature. Feature modules should configure their own DI wherever possible for
+feature-specific (non-global) services.
```

Comment:
suggestion(non-blocking): preferably in standalone components or using `provideIn: 'root'` and not in the actual feature module

Reply:
Why not? If the component is standalone, then by definition it doesn't have a module, so advice about a feature module doesn't apply. And if you have a service that is only used in a feature module, I assume you'd want to register it in the more specific location rather than in root. Does this go to your earlier concern about accidentally instantiating duplicate services?

Reply:
> If the component is standalone, then by definition it doesn't have a module, so advice about a feature module doesn't apply. 

Good point, I read `feature module` as just a folder in an organize-by-feature structure. I think feature modules could (and should) be replaced by standalone component, but that discussion might be outside the scope of this PR comment :D

> Does this go to your earlier concern about accidentally instantiating duplicate services?

Yes, and other things. It depends a little bit on how we define a "feature". If `feature => 1 exported top-level component` then we are basically talking about standalone components. 

In the case where 1 feature module exports multiple components then the module becomes a bit rigid. It's a big blob containing all of the dependencies that any of it's child components *might* need. It makes it less obvious which component needs which service/pipe/etc. and also affects tree-shaking negatively. The lifetime of the service also becomes much easier to reason about, since it's tied to an actual DOM component, instead of a module that is somewhat magically instantiated by Angular in the background. Service tied to components also have access to the activated route in a way that service in modules never will (though this use-case might be a bit of an edge case).

Reply:
I added a mention of standalone components when discussing the web vault client, otherwise I suggest this could be done in a subsequent revision of this page - which would definitely be welcome. It might require some discussion/consensus on stricter rules around feature modules though. I agree in many cases they can be replaced with standalone, but it's not what we do today.